Transform into vendor-neutral upstream project with zero hardcoded values

rrasouli opened the pull request:
## Summary

BYOH Provisioner for deploying Windows worker nodes to OpenShift clusters across multiple cloud platforms.

## Overview

This tool provisions Windows worker nodes as BYOH (Bring Your Own Host) for OpenShift clusters. It supports AWS, Azure, GCP, vSphere, Nutanix, and bare metal environments with automated credential management and infrastructure discovery from existing Linux worker nodes.

## What This Tool Does

### Multi-Cloud Windows Node Provisioning

Deploy Windows Server 2019/2022 nodes across platforms:
- AWS
- Azure
- GCP
- vSphere
- Nutanix
- Bare metal (platform "none")

### Automated Infrastructure Discovery

The tool extracts infrastructure configuration from Linux worker nodes in the cluster:

```bash
# AWS: Region, VPC, subnet, security groups
linux_machine_spec=$(oc get machines -n openshift-machine-api \
  -l machine.openshift.io/cluster-api-machine-role=worker \
  -o=jsonpath='{.items[0].spec}')
region=$(echo "$linux_machine_spec" | jq -r '.providerSpec.value.placement.region')
vpc=$(echo "$linux_machine_spec" | jq -r '.providerSpec.value.vpc')

# Azure: VNet, subnet
vnet=$(echo "$linux_machine_spec" | jq -r '.providerSpec.value.vnet')
subnet=$(echo "$linux_machine_spec" | jq -r '.providerSpec.value.subnet')

# vSphere: Datacenter, datastore, network
datacenter=$(echo "$linux_machine_spec" | jq -r '.workspace.datacenter')
datastore=$(echo "$linux_machine_spec" | jq -r '.workspace.datastore')
network=$(echo "$linux_machine_spec" | jq -r '.network.devices[0].networkName')
```

**Benefits:**
- No manual infrastructure parameter entry
- Network configuration matches Linux workers
- Works immediately with any OpenShift cluster
- No Windows MachineSet prerequisite

### Automated Credential Management

**Auto-Generated Passwords:**
```bash
# Cryptographically secure 18-character random password
function generate_random_password() {
    echo "$(dd if=/dev/urandom bs=1 count=101 2>/dev/null | tr -dc 'a-z0-9A-Z' | head -c 18)"
}
```

**Auto-Extracted SSH Keys:**
```bash
# Extract from cloud-private-key secret in WMCO namespace
function get_ssh_public_key_from_secret() {
    local wmco_namespace=$(oc get deployment --all-namespaces \
      -o=jsonpath="{.items[?(@.metadata.name=='windows-machine-config-operator')].metadata.namespace}")

    local private_key=$(oc get secret cloud-private-key -n "$wmco_namespace" \
      -o jsonpath='{.data.private-key\.pem}' | base64 -d)

    ssh-keygen -y -f <(echo "$private_key")
}
```

**Platform-Specific Credentials:**
- Azure: Extracts all 6 variables from cluster secrets (ARM_CLIENT_ID, ARM_CLIENT_SECRET, ARM_SUBSCRIPTION_ID, ARM_TENANT_ID, ARM_RESOURCE_PREFIX, ARM_RESOURCEGROUP)
- AWS: Supports multiple credential methods (profile + shared file, direct keys, default credentials)
- GCP: Service account from cluster secrets
- vSphere: vCenter credentials from cluster secrets
- Nutanix: Prism Central credentials from cluster secrets

### Generic Windows Bootstrap Template

Single cross-platform bootstrap script at `lib/windows-vm-bootstrap.tf`:

```powershell
<powershell>
# Configure Administrator account for OpenSSH authentication
# OpenSSH on Windows requires valid password to generate security token (LogonUser API)
$UserAccount = Get-LocalUser -Name "${var.admin_username}" -ErrorAction SilentlyContinue
if ($UserAccount -ne $null) {
    $password = ConvertTo-SecureString "${var.admin_password}" -AsPlainText -Force
    $UserAccount | Set-LocalUser -Password $password -PasswordNeverExpires $true
    if (!$UserAccount.Enabled) {
        $UserAccount | Enable-LocalUser
    }
}

# Setup SSH authorized keys
$authorizedKeyConf = "$env:ProgramData\ssh\administrators_authorized_keys"
Write-Output "${var.ssh_public_key}" | Out-File -FilePath $authorizedKeyConf -Encoding ascii

# Install and configure OpenSSH Server
Add-WindowsCapability -Online -Name OpenSSH.Server~~~~0.0.1.0
Set-Service -Name sshd -StartupType 'Automatic'
Start-Service sshd

# Configure firewall
New-NetFirewallRule -DisplayName "ContainerLogsPort" -LocalPort ${var.container_logs_port} -Enabled True -Direction Inbound -Protocol TCP -Action Allow -EdgeTraversalPolicy Allow
</powershell>
```

Platform-specific directories contain symlinks to this generic template:
- `aws/windows-vm-bootstrap.tf` → `../lib/windows-vm-bootstrap.tf`
- `gcp/windows-vm-bootstrap.tf` → `../lib/windows-vm-bootstrap.tf`
- `azure/windows-vm-bootstrap.tf` → `../lib/windows-vm-bootstrap.tf`
- `nutanix/windows-vm-bootstrap.tf` → `../lib/windows-vm-bootstrap.tf`

### Image Selection Strategy

Per-platform priority ordering:

| Platform | Priority |
|----------|----------|
| **AWS** | User Config → AWS API (version-specific) → MachineSet |
| **Azure** | User Config → MachineSet → Default SKU (latest) |
| **GCP** | Image family (always latest) |
| **vSphere** | User Config → MachineSet → Error |
| **Nutanix** | User Config → MachineSet → Error |

**AWS Example:**
```bash
# Priority 1: User override
windows_ami=$(get_config "AWS_WINDOWS_AMI" "")

# Priority 2: AWS API query (version-specific)
if [[ -z "$windows_ami" ]] && command -v aws &> /dev/null; then
    image_pattern="Windows_Server-${win_version}-English-Full-Base"
    windows_ami=$(aws ec2 describe-images \
        --filters "Name=name,Values=${image_pattern}*" \
        --region "${region}" \
        --query 'sort_by(Images, &CreationDate)[-1].[ImageId]' \
        --output text)
fi

# Priority 3: MachineSet fallback
if [[ -z "$windows_ami" ]]; then
    windows_ami=$(oc get machineset -n openshift-machine-api ...)
fi
```

### Configuration System

Configuration priority (highest to lowest):
1. Environment variables
2. User config file (`~/.config/byoh-provisioner/config`)
3. Project config file (`./configs/defaults.conf`)
4. Built-in defaults

**Implementation:**
```bash
# lib/config.sh - Only export if not already set in environment
if [[ -z "${!key:-}" ]]; then
    export "$key=$value"
fi
```

### Modular Architecture

```
terraform-windows-provisioner/
├── byoh.sh                      # Main entry point
├── lib/                         # Library modules
│   ├── config.sh               # Configuration loading (bash 3.x compatible)
│   ├── credentials.sh          # Credential management
│   ├── platform.sh             # Platform detection & tfvars generation
│   ├── terraform.sh            # Terraform operations (cp -LR for symlinks)
│   ├── validation.sh           # Input validation
│   └── windows-vm-bootstrap.tf # Generic Windows bootstrap (all platforms)
├── configs/                    # Configuration files
│   ├── defaults.conf           # Default values
│   └── examples/               # Platform-specific examples
└── <platform>/                 # Platform-specific Terraform
    ├── aws/
    ├── azure/
    ├── gcp/
    ├── vsphere/
    ├── nutanix/
    └── none/
```

### Key Configuration Variables

| Variable | Description | Default |
|----------|-------------|---------|
| `WMCO_NAMESPACE` | WMCO deployment namespace | Auto-detected |
| `WMCO_IDENTIFIER_TYPE` | ConfigMap identifier: `ip` or `dns` | `ip` |
| `WINC_ADMIN_USERNAME` | Windows admin username | Azure=`capi`, Others=`Administrator` |
| `WINC_ADMIN_PASSWORD` | Windows password | Auto-generated |
| `WINC_SSH_PUBLIC_KEY` | SSH public key | Auto-extracted |
| `AWS_WINDOWS_AMI` | AWS AMI override | Auto-detected |
| `AZURE_WINDOWS_SKU` | Azure image SKU | `{version}-Datacenter-smalldisk` |
| `VSPHERE_WINDOWS_TEMPLATE` | vSphere template | From MachineSet |
| `NUTANIX_WINDOWS_IMAGE` | Nutanix image | From MachineSet |

## Usage Examples

### Basic Deployment
```bash
./byoh.sh apply mywindows 2
```

### Specific Windows Version
```bash
./byoh.sh apply mywindows 2 '' 2019
```

### Custom Configuration
```bash
export WINC_ADMIN_PASSWORD="MyPassword123"
export AWS_WINDOWS_AMI="ami-0abcdef1234567890"
./byoh.sh apply production-win 4
```

### Cleanup
```bash
./byoh.sh destroy mywindows 2
```

## Technical Details

### Files in This PR

| File | Purpose |
|------|---------|
| `byoh.sh` | Main entry point, command parsing |
| `lib/config.sh` | Configuration loading (bash 3.x compatible) |
| `lib/credentials.sh` | Credential management, auto-generation |
| `lib/platform.sh` | Platform detection, infrastructure discovery, AWS version selection |
| `lib/terraform.sh` | Terraform operations, ConfigMap management (cp -LR for symlinks) |
| `lib/validation.sh` | Input validation |
| `lib/windows-vm-bootstrap.tf` | Generic Windows bootstrap template |
| `aws/main.tf` | AWS Terraform configuration |
| `aws/variables.tf` | AWS variables |
| `azure/main.tf` | Azure Terraform configuration |
| `azure/variables.tf` | Azure variables |
| `gcp/main.tf` | GCP Terraform configuration |
| `gcp/variables.tf` | GCP variables (includes container_logs_port) |
| `vsphere/main.tf` | vSphere Terraform configuration |
| `vsphere/variables.tf` | vSphere variables |
| `nutanix/main.tf` | Nutanix Terraform configuration |
| `nutanix/variables.tf` | Nutanix variables (includes container_logs_port) |
| `none/main.tf` | Platform "none" Terraform configuration |
| `none/variables.tf` | Platform "none" variables |
| `configs/defaults.conf` | Default configuration values |
| `configs/examples/*.conf.example` | Platform-specific config examples |
| `README.md` | Comprehensive documentation |

### Bash 3.x Compatibility

RHEL 8 uses bash 3.2, so the code avoids bash 4.0+ features:

```bash
# Avoid associative arrays (bash 4.0+)
# Use space-delimited strings with pattern matching instead:
ENV_VARS_BEFORE_CONFIG=""
for var in "${config_vars[@]}"; do
    if [[ -n "${!var:-}" ]]; then
        ENV_VARS_BEFORE_CONFIG="${ENV_VARS_BEFORE_CONFIG} ${var}"
    fi
done

# Check membership
if [[ ! " ${ENV_VARS_BEFORE_CONFIG} " =~ " ${key} " ]]; then
    export "$key=$value"
fi
```

### AWS Version Selection

Respects requested Windows version with proper priority:

```bash
function write_aws_tfvars() {
    local win_version="${4:-2022}"
    local skip_ami_lookup="${5:-false}"  # Optimize destroy operations

    if [[ "$skip_ami_lookup" == "true" ]]; then
        windows_ami="ami-dummy-not-used-for-destroy"
    else
        # Version-specific query with 30s timeout
        image_pattern="Windows_Server-${win_version}-English-Full-Base"
        windows_ami=$(timeout 30s aws ec2 describe-images ...)
    fi
}
```

### Symlink Handling

Terraform doesn't follow symlinks, so use `cp -LR` to dereference:

```bash
# lib/terraform.sh
cp -LR "${script_dir}/${platform}/." "$templates_dir"
```

## Platform Coverage

- AWS (tested)
- Azure (tested)
- GCP (tested)
- vSphere (tested)
- Nutanix (tested)
- Platform "none" / bare metal (tested)

## Integration with OpenShift CI

This tool integrates with Prow CI via step-registry. See `ci-operator/step-registry/windows/byoh/` in openshift/release repository.

## Benefits

**For Users:**
- No manual credential generation
- No manual infrastructure parameter entry
- Works immediately with existing clusters
- Single command deployment

**For CI/CD:**
- Multi-platform support
- Environment variable configuration
- No interactive prompts
- Clean Terraform state management

**For Operations:**
- Automated infrastructure discovery
- Platform auto-detection
- Graceful error handling
- Transparent configuration priority

## Testing

Validated on:
- AWS: IPI clusters, Windows 2019/2022, credential methods
- Azure: IPI clusters, Windows 2019/2022, image SKUs
- GCP: IPI clusters, Windows 2019/2022, service accounts
- vSphere: Golden image templates, Windows 2019/2022
- Nutanix: Golden images, Windows 2019/2022
- Platform "none": Bare metal with AWS infrastructure

## Reviewers

Please review:
- Modular architecture and separation of concerns
- Generic Windows bootstrap template approach
- Infrastructure discovery from Linux workers
- Credential auto-generation security
- AWS version selection priority
- Bash 3.x compatibility
- Configuration priority implementation
- Error handling and user messaging
- Documentation completeness

rrasouli committed:
1. BYOH provisioner for multi-cloud Windows node deployment

- Generic, configurable tool for provisioning Windows worker nodes
- Multi-cloud support: AWS, Azure, GCP, vSphere, Nutanix, bare metal
- Intelligent infrastructure discovery from Linux Machine API
- Automated credential management with auto-generated passwords
- Platform-specific image selection with smart fallback
- Zero hardcoded values, fully parameterized configuration

2. configurable WMCO namespace and DNS-based CM support

Features:
- Add WMCO_NAMESPACE configuration with auto-detection fallback
  - Auto-detects namespace by finding windows-machine-config-operator deployment
  - Supports custom namespace override via configuration
  - Validates configured namespace exists before use

- Add WMCO_IDENTIFIER_TYPE for flexible ConfigMap identifiers
  - Supports 'ip' (default) for IP addresses
  - Supports 'dns' for DNS hostnames/FQDNs
  - Graceful fallback from DNS to IP if Terraform output unavailable
  - Useful for automation tests requiring DNS-based identification

Implementation:
- New get_wmco_namespace() function in lib/terraform.sh
  - Checks configured namespace first
  lib

- Updated get_ssh_public_key_from_secret() in lib/credentials.sh
  - Uses get_wmco_namespace() for consistency

Terraform outputs:
- Add instance_hostname output to all platforms:
  - AWS: private_dns
  - Azure: computer_name
  - GCP: instance name
  - vSphere: VM name
  - Nutanix: VM name
  - None (baremetal): private_dns

Configuration:
- WMCO_NAMESPACE defaults to 'openshift-windows-machine-config-operator'
- WMCO_IDENTIFIER_TYPE defaults to 'ip'
- Both configurable via environment, user config, or project defaults
- Full configuration priority chain support

Documentation:
- Updated README.md with configuration variables and examples
- Updated PR_DESCRIPTION.md with implementation details
- Added user config template examples in lib/config.sh

Benefits:
- Works with custom WMCO namespace deployments
- Supports automation tests requiring DNS identifiers
- 100% backward compatible (defaults unchanged)
- Graceful error handling and fallbacks
- Clear logging for troubleshooting

mburke5678 commented on README.md:
```diff
@@ -0,0 +1,318 @@
+# BYOH Provisioner - Bring Your Own Host for Windows Nodes
+
+A generic, configurable tool for provisioning and managing Windows worker nodes across multiple cloud platforms. This project provides automated BYOH (Bring Your Own Host) Windows node deployment for Kubernetes/OpenShift clusters with support for AWS, Azure, GCP, vSphere, Nutanix, and bare metal environments.
```
This should be a full sentence:
```suggestion
The BYOH Provisioner is a generic, configurable tool for provisioning and managing Windows worker nodes across multiple cloud platforms. This project provides automated BYOH (Bring Your Own Host) Windows node deployment for Kubernetes/OpenShift clusters with support for AWS, Azure, GCP, vSphere, Nutanix, and bare metal environments.
```

mburke5678 commented on README.md:
````diff
@@ -0,0 +1,318 @@
+
+### 1. Install the Tool
+
+```bash
+git clone https://github.com/your-org/terraform-windows-provisioner.git
+cd terraform-windows-provisioner
+chmod +x byoh.sh
+```
````
So that users know to replace `your-org` with the name of their org.
```suggestion
```bash
git clone https://github.com/<your-org>/terraform-windows-provisioner.git
cd terraform-windows-provisioner
chmod +x byoh.sh
```
```

mburke5678 commented on README.md:
````diff
@@ -0,0 +1,318 @@
+./byoh.sh apply mywindows 4 '' 2019
+```
+
+### 4. Destroy When Done
````
```suggestion
### 4. Destroy Windows Nodes When Done
```

mburke5678 commented on README.md:
````diff
@@ -0,0 +1,318 @@
+
+```bash
+# Create instances (default: 2x Windows Server 2022)
+./byoh.sh apply [NAME] [NUM_WORKERS] [FOLDER_SUFFIX] [WINDOWS_VERSION]
````
```suggestion
# Create instances (default: two Windows Server 2022 instances named `byoh-winc-*`)
./byoh.sh apply [NAME] [NUM_WORKERS] [FOLDER_SUFFIX] [WINDOWS_VERSION]
```

mburke5678 commented on README.md:
````diff
@@ -0,0 +1,318 @@
+
+# Show what would be deployed without creating
+./byoh.sh arguments myapp 2
+```
````
I think this table and examples should go before [### Basic Commands](https://github.com/openshift/terraform-windows-provisioner/pull/6/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R76) so that users will know what the parameters in the command are. It would also eliminate the need for the  **(default: 2x Windows Server 2022)** in [# Create instances (default: 2x Windows Server 2022)](https://github.com/openshift/terraform-windows-provisioner/pull/6/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R79).
Unless this is the way things are done in documents like this.

mburke5678 commented on README.md:
```diff
@@ -0,0 +1,318 @@
+
+## Support
+
+- **Issues**: Report bugs or request features via GitHub Issues
```
Maybe a link to the repo issues page? Something like:
```suggestion
- **Issues**: Report bugs or request features via [GitHub Issues](https://github.com/openshift/terraform-windows-provisioner/issues)
```

mburke5678 commented on README.md:
```diff
@@ -0,0 +1,318 @@
+
+## Security
+
+For security issues, please see [SECURITY.md](SECURITY.md).
```
This file, SECURITY.md, isn't present that I can see; as such the link doesn't work.

mburke5678 commented on README.md:
```diff
@@ -0,0 +1,318 @@
+
+---
+
+**Note**: This is a generic, vendor-neutral project with no hardcoded values. All configuration is customizable via configuration files or environment variables.
```
Could this go at the beginning of the file, [right after the introductory paragraph](https://github.com/openshift/terraform-windows-provisioner/pull/6/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R4)?

rrasouli replied:
Done

mburke5678 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,223 @@
+
+### Pull Requests
+
+1. **Fork the repository** and create your branch from `main`
```
```suggestion
1. **Fork the repository** and create your branch from `main`.
```

mburke5678 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,223 @@
+2. **Make your changes**:
+   - Follow the coding standards (see below)
+   - Add or update tests as needed
+   - Update documentation as needed
```
List items that are full sentences should have a period. 
```suggestion
   - Follow the coding standards (see below).
   - Add or update tests as needed.
   - Update documentation as needed.
```

mburke5678 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,223 @@
+3. **Test your changes**:
+   - Test on at least one cloud platform
+   - Ensure existing functionality still works
+   - Run shellcheck on bash scripts
```
List items that are full sentences should have a period. 
```suggestion
   - Test on at least one cloud platform.
   - Ensure existing functionality still works.
   - Run shellcheck on bash scripts.
```

mburke5678 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,223 @@
+   - Run shellcheck on bash scripts
+4. **Commit your changes**:
+   - Use clear, descriptive commit messages
+   - Reference any related issues
```
List items that are full sentences should have a period. 
```suggestion
   - Use clear, descriptive commit messages.
   - Reference any related issues.
```

mburke5678 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,223 @@
+4. **Commit your changes**:
+   - Use clear, descriptive commit messages
+   - Reference any related issues
+5. **Push to your fork** and submit a pull request
```
```suggestion
5. **Push to your fork** and submit a pull request.
```

mburke5678 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,223 @@
+- Use functions for reusable code
+- Handle errors gracefully with informative messages
+- Use `local` for function variables
+- Quote variables: `"${variable}"`
```
```suggestion
- Use `set -euo pipefail` at the beginning.
- Use `shellcheck` and address all warnings.
- Use meaningful variable names.
- Add comments for complex logic.
- Use functions for reusable code.
- Handle errors gracefully with informative messages.
- Use `local` for function variables.
- Quote variables: `"${variable}"`.
```

mburke5678 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,223 @@
+- Add descriptions to all variables
+- Mark sensitive variables appropriately
+- Use consistent naming conventions
+- Add comments for complex configurations
```
```suggestion
- Use `terraform fmt` for formatting.
- Use meaningful resource names.
- Add descriptions to all variables.
- Mark sensitive variables appropriately.
- Use consistent naming conventions.
- Add comments for complex configurations.
```

mburke5678 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,223 @@
+- Keep README.md up to date
+- Document all configuration variables
+- Provide examples for common use cases
+- Update CHANGELOG.md for notable changes
```
```suggestion
- Keep README.md up to date.
- Document all configuration variables.
- Provide examples for common use cases.
- Update CHANGELOG.md for notable changes.
```

mburke5678 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,223 @@
+3. Update `lib/platform.sh`:
+   - Add platform to `SUPPORTED_PLATFORMS`
+   - Add `get_<platform>_terraform_args()` function
+   - Add `get_user_name()` case
```
```suggestion
   - Add platform to `SUPPORTED_PLATFORMS`.
   - Add `get_<platform>_terraform_args()` function.
   - Add `get_user_name()` case.
```

mburke5678 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,223 @@
+   - Add `get_<platform>_terraform_args()` function
+   - Add `get_user_name()` case
+4. Update `lib/credentials.sh`:
+   - Add `export_<platform>_credentials()` function
```
```suggestion
   - Add `export_<platform>_credentials()` function.
```

mburke5678 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,223 @@
+3. Add variable to relevant Terraform `variables.tf`
+4. Update `lib/platform.sh` to pass the variable
+5. Document in README.md
+6. Add example usage
```
```suggestion
1. Add to `configs/defaults.conf` with default value.
2. Update `configs/examples/defaults.conf.example`.
3. Add variable to relevant Terraform `variables.tf`.
4. Update `lib/platform.sh` to pass the variable.
5. Document in README.md.
6. Add example usage.
```

mburke5678 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,223 @@
+- Use present tense ("Add feature" not "Added feature")
+- Use imperative mood ("Move cursor to..." not "Moves cursor to...")
+- Limit first line to 72 characters
+- Reference issues and pull requests liberally
```
```suggestion
- Use present tense ("Add feature" not "Added feature").
- Use imperative mood ("Move cursor to..." not "Moves cursor to...").
- Limit first line to 72 characters.
- Reference issues and pull requests liberally.
```

mburke5678 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,223 @@
+
+Fix credential loading for vSphere platform
+
+Update documentation for configuration system
```
```suggestion
Add support for custom Windows image versions in Azure.

Parameterize all hardcoded values in AWS Terraform.

Fix credential loading for vSphere platform.

Update documentation for configuration system.
```

mburke5678 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,223 @@
+1. Update VERSION file
+2. Update CHANGELOG.md
+3. Create release tag
+4. Create GitHub release with notes
```
```suggestion
1. Update the VERSION file.
2. Update the CHANGELOG.md.
3. Create the release tag.
4. Create the GitHub release with notes.
```

mburke5678 commented on docs/PROW_CI_INTEGRATION.md:
```diff
@@ -0,0 +1,168 @@
+3. Extracts Windows SSH key from WMCO `cloud-private-key` secret
+4. Clones terraform-windows-provisioner repository
+5. Runs provisioning via `byoh.sh apply`
+6. Exports instance information to `${SHARED_DIR}` for WMCO tests
```
```suggestion
1. Detects the platform from the cluster infrastructure.
2. Loads the cloud credentials from the `CLUSTER_PROFILE_DIR` directory.
3. Extracts the Windows SSH key from the WMCO `cloud-private-key` secret.
4. Clones the terraform-windows-provisioner repository.
5. Runs provisioning via the `byoh.sh apply` command.
6. Exports instance information to the `${SHARED_DIR}` for WMCO tests.
```

mburke5678 commented on docs/PROW_CI_INTEGRATION.md:
```diff
@@ -0,0 +1,168 @@
+
+1. Reads work directory from provision step
+2. Runs cleanup via `byoh.sh destroy`
+3. Removes temporary files
```
```suggestion
1. Reads the work directory from the provision step.
2. Runs the cleanup via the `byoh.sh destroy` command.
3. Removes the temporary files.
```

mburke5678 commented on docs/PROW_CI_INTEGRATION.md:
```diff
@@ -0,0 +1,168 @@
+
+1. Make changes to step-registry components in `openshift/release`
+2. Test with `/pj-rehearse` command
+3. Submit PR to `openshift/release` repository
```
```suggestion
1. Make changes to the step-registry components in the `openshift/release` repository.
2. Test with the `/pj-rehearse` command.
3. Submit PR to `openshift/release` repository.
```

mburke5678 commented on docs/PROW_CI_INTEGRATION.md:
```diff
@@ -0,0 +1,168 @@
+For changes to the provisioner itself:
+
+1. Make changes in this repository
+2. The step-registry will automatically use the latest version from `main` branch
```
```suggestion
1. Make changes in this repository.
2. The step-registry will automatically use the latest version from `main` branch.
```

mburke5678 commented on docs/PROW_CI_INTEGRATION.md:
```diff
@@ -0,0 +1,168 @@
+
+- **Step-Registry Issues**: Create issues in [openshift/release](https://github.com/openshift/release/issues)
+- **Provisioner Issues**: Create issues in [terraform-windows-provisioner](https://github.com/openshift/terraform-windows-provisioner/issues)
+- **Slack**: `#windows-containers` or `#forum-ocp-testplatform`
```
```suggestion
- **Step-Registry Issues**: Create issues in [openshift/release](https://github.com/openshift/release/issues).
- **Provisioner Issues**: Create issues in [terraform-windows-provisioner].(https://github.com/openshift/terraform-windows-provisioner/issues)
- **Slack**: `#windows-containers` or `#forum-ocp-testplatform`
```

mburke5678 commented on SECURITY.md:
```diff
@@ -0,0 +1,40 @@
+2. **Secrets**: Use Kubernetes secrets for sensitive data
+3. **Access Control**: Limit access to service accounts and API keys
+4. **Network Security**: Follow your organization's network security policies
+5. **Updates**: Keep Terraform and dependencies up to date
```
```suggestion
1. **Credentials**: Never commit credentials to version control.
2. **Secrets**: Use Kubernetes secrets for sensitive data.
3. **Access Control**: Limit access to service accounts and API keys.
4. **Network Security**: Follow your organization's network security policies.
5. **Updates**: Keep Terraform and dependencies up to date.
```

mburke5678 commented on SECURITY.md:
```diff
@@ -0,0 +1,40 @@
+
+- Cloud provider credentials are required and should be managed securely
+- Windows admin passwords are used for provisioning - use strong passwords
+- SSH keys are extracted from cluster secrets - ensure proper RBAC is configured
```
```suggestion
- Cloud provider credentials are required and should be managed securely.
- Windows admin passwords are used for provisioning - use strong passwords.
- SSH keys are extracted from cluster secrets - ensure proper RBAC is configured.
```

mburke5678 reviewed (changes requested):
A few suggestions, mostly grammar things, for SECURITY.md, CONTRIBUTING.md, and PROW_CI_INTEGRATION.md.

mburke5678 reviewed (commented):
Previously reviewed

rrasouli commented:
Some issues found in GCP, Azure WMCO configuration with SSH

weinliu commented:
Thank you for this comprehensive PR! The overall direction and features are excellent, but I have some concerns that need to be addressed before merging.
1. **Password generation doesn't meet Windows requirements** in `lib/credentials.sh`, shall we update it?
2. Do we need to add error handling in SSH key extraction function?

rrasouli commented:
> Thank you for this comprehensive PR! The overall direction and features are excellent, but I have some concerns that need to be addressed before merging.
> 
> 1. **Password generation doesn't meet Windows requirements** in `lib/credentials.sh`, shall we update it?
> 2. Do we need to add error handling in SSH key extraction function?

1. which Windows requirements? I don't undertand
2. I think we have added such errors - it is nice to have, perhaps we can add a debug mode in the future

weinliu commented:
1. `lib/credentials.sh` Missing special characters. I think it's ok, since the windows are used for testing
Other part looks good.

Waiting for the fix for 

> Some issues found in GCP, Azure WMCO configuration with SSH

rrasouli committed:
Fix Windows BYOH SSH authentication failures on GCP, Azure

This commit resolves critical SSH connectivity issues that prevented WMCO
from configuring Windows BYOH nodes. The root cause was OpenSSH on Windows
failing to generate user tokens for authentication.

Critical Fixes:
 - GCP/Azure: Always set Administrator/user password with PasswordNeverExpires
   to ensure OpenSSH can generate authentication tokens, regardless of account
   enabled state. This was THE root cause of "unable to connect to Windows VM"
   timeouts.

 - GCP/Azure: Create scheduled task to start ssh-agent and sshd services at
   every boot. Services configured during sysprep don't reliably persist after
   Windows reboot (RC bug).

 - GCP/Azure: Insert UseDNS directive before Match block in sshd_config to
   prevent syntax errors that stop sshd from starting.

 - GCP/Azure: Ensure ssh-agent service starts before sshd (Windows requirement).

Infrastructure Improvements:
 - lib/terraform.sh: Use ConfigMap PATCH instead of DELETE+CREATE to preserve
   other BYOH node entries in multi-node environments.

 - byoh.sh: Add SKIP_CONFIGMAP_CREATION support for manual ConfigMap management.
   Regenerate terraform.auto.tfvars before destroy to prevent variable errors.

 - lib/credentials.sh: Add SSH key validation and improved password generation
   meeting Azure complexity requirements.

 - lib/platform.sh: Extract Azure image version from existing MachineSet instead
   of hardcoding "latest" to avoid buggy releases.

 - lib/config.sh: Improve config file parsing to handle whitespace robustly.

 - azure/main.tf: Add 120s wait before VM extension execution to prevent race
   conditions during VM boot.

Tested on:
 - GCP: windows-2022-core ✓
   - Azure: windows-2019-datacenter ✓
   - vSphere: ✓

jrvaldes commented on README.md:
```diff
@@ -7,6 +7,7 @@ The BYOH Provisioner is a generic, configurable tool for provisioning and managi
 ## Features
 
 - **Multi-Cloud Support**: Deploy Windows nodes on AWS, Azure, GCP, vSphere, Nutanix, or bare metal
+- **Production-Ready SSH Authentication**: Critical fixes for Windows OpenSSH token generation, ensuring WMCO can reliably connect to BYOH nodes
```
I do not recommend using terms like `Production-Ready` in an internal script aimed for CI.

rrasouli replied:
Thanks for the comment, removed all "Production Ready" in this project

jrvaldes commented on azure/windows-vm-bootstrap.tf:
```diff
@@ -39,22 +69,18 @@ $systemRule = New-Object system.security.accesscontrol.filesystemaccessrule("SYS
-# success
+
+# Create scheduled task to ensure SSH services start on every boot
+$Action = New-ScheduledTaskAction -Execute 'powershell.exe' -Argument '-NoProfile -WindowStyle Hidden -Command "Start-Service ssh-agent; Start-Service sshd"'
```
> Set-Service -Name sshd -StartupType 'Automatic'

StartupType was set to 'Automatic' , this may be redundant

rrasouli replied:
Not redundant, both are necessary (belt-and-suspenders approach):

  - Lines 36-37: Set-Service -StartupType 'Automatic' - The "proper" way, what SHOULD work
  - Lines 76-80: Scheduled Task - What we NEED because services configured during sysprep don't reliably persist after reboot
**The Problem**: On GCP/Azure, services configured during sysprep-specialize/VM extension phases don't always respect the StartupType 'Automatic'
setting after the automatic Windows reboot:
The  StartupType was set, but sshd didn't start after reboot.

Hence, Scheduled task running as SYSTEM at boot is more reliable. This was the Root cause bug fix.

jrvaldes commented on azure/windows-vm-bootstrap.tf:
```diff
@@ -5,6 +5,21 @@ data "template_file" "windows-userdata" {
 Rename-Computer -NewName "${var.winc_instance_name}-${count.index}" -Force;
+
+# Configure user account for OpenSSH authentication
+# Azure uses 'capi' user, but ensure password is set for token generation
```
> token generation

what is the "token generation" in this context?

rrasouli replied:
**"User token generation"** = Windows Security Token (LogonUser token)

  When OpenSSH on Windows authenticates a user via SSH, it needs to:
  1. Call LogonUser() Windows API to get a security token for that user
  2. Use that token to impersonate the user for running commands
  3. This requires the user account to have a valid password configured

  Without password set with PasswordNeverExpires, OpenSSH fails with:
  `sshd: fatal: ga_init, unable to resolve user administrator`
  `sshd: error: get_user_token - unable to generate token on 2nd attempt for user administrator`

sebsoto commented:
/lgtm

jrvaldes commented:
/lgtm

rrasouli commented:
/approve

rrasouli commented:
/approve

rrasouli commented:
@sdodson Could you please merge this PR? It has both the `lgtm` and
`approved` labels, and all requirements are met.
The repository doesn't have Tide auto-merge configured, so it needs manual merging. Since you merged PR #1, you have the necessar permissions.
Thank you!

rrasouli commented:
@JimCann Could you please help with two things?

1. **Merge this PR**: It has both the `lgtm` and `approved` labels, and all requirements are met. The repository doesn't have Tide auto-merge configured, so it needs manual merging.

2. **Grant me write permissions**: I'm listed as an approver in the OWNERS file, but I only have READ permissions on the GitHub repository itself, so I can't merge PRs. Could you grant me write access so I can manage PRs going forward?

Thank you!
